### PR TITLE
Help tool for Wx backends

### PR DESCRIPTION
**Pull request by DietmarSchwertberger**
## PR Summary
wx implementation of the ToolManager Help tool (see #11045)

Re-factored part of `ToolHelpBase._get_help_text` and `_get_help_html` into `_get_help_entries` as this implementation needs a list of entries.

wxhelptool

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/backends/backend_wx.py [hunk 1]
```diff
@@ -1826,9 +1826,49 @@ def remove_rubberband(self, dc=None):
             self._rect = None
 
 
+class _TableDialog(wx.Dialog):
+    def __init__(self, parent, help, title="Help"):
```
Comment (timhoffm) [inline lib/matplotlib/backends/backend_wx.py]: `help` is built-in function. While technically you can use that variable name, I would not advise to do so. Also, it's slightly imprecise. Maybe rename to `help_entries`.

File: lib/matplotlib/backends/backend_wx.py [hunk 2]
```diff
@@ -1826,9 +1826,49 @@ def remove_rubberband(self, dc=None):
...
+        self.Bind(wx.EVT_CLOSE, self.OnClose)
+
+    def OnClose(self, evt):
+        self.DestroyLater()
+        evt.Skip()
+
+
+class HelpWx(backend_tools.ToolHelpBase):
+    def trigger(self, *args):
+        help = [("Action","Shortcuts", "Description")]
```
Comment (timhoffm) [inline lib/matplotlib/backends/backend_wx.py]: `help_entries`?
Also a missing space.

File: lib/matplotlib/backend_tools.py [hunk 3]
```diff
@@ -1039,30 +1039,31 @@ def _format_tool_keymap(self, name):
...
-                continue
-            rows.append(fmt.format(
-                name, self._format_tool_keymap(name), tool.description))
+        rows += [fmt.format(*row) for row in self._get_help_entries()]
         return ("<style>td {padding: 0px 4px}</style>"
                 "<table><thead>" + rows[0] + "</thead>"
                 "<tbody>".join(rows[1:]) + "</tbody></table>")
 
 
+
```
Comment (timhoffm) [inline lib/matplotlib/backend_tools.py]: Remove. PEP8 will complain about too many empty lines.

**Comment by DietmarSchwertberger** [comment]
Thanks, I probably should install a PEP 8 checker...
Yes, `help_entries` is much better.
I have modified the code to avoid multiple dialogs being created shown.
I hope it is OK, to keep a reference to the dialog as class variable and reset it from the dialog when it's closed and destroyed:
```
class _HelpDialog(wx.Dialog):
    def OnClose(self, evt):
        HelpWx.dlg = None
        self.DestroyLater()
        ...

class HelpWx(backend_tools.ToolHelpBase):
    dlg = None
```

File: lib/matplotlib/backends/backend_wx.py [hunk 4]
```diff
@@ -1826,9 +1826,56 @@ def remove_rubberband(self, dc=None):
...
+        OK = wx.Button(self, wx.ID_OK)
+        sizer.Add(OK, 0, wx.ALIGN_CENTER_HORIZONTAL | wx.ALL, 8)
+        self.SetSizer(sizer)
+        sizer.Fit(self)
+        self.Layout()
+        self.Bind(wx.EVT_CLOSE, self.OnClose)
+        OK.Bind(wx.EVT_BUTTON, self.OnClose)
+
+    def OnClose(self, evt):
+        HelpWx.dlg = None
```
Comment (anntzer) [inline lib/matplotlib/backends/backend_wx.py]: That'll reset the class variable, but in HelpWx you're setting the instance variable.

**Comment by timhoffm** [comment]
Don't know how these things are usually handled in wx. It's not particularly nice, but ok for me. Maybe add a short comment that makes this dependency handling clear.

File: lib/matplotlib/backends/backend_wx.py (hunk 4 above)
Comment (DietmarSchwertberger) [reply lib/matplotlib/backends/backend_wx.py]: Oh, yes, you're right.
I don't want to pass and store a reference to the HelpWx instance in the dialog.
On the other hand, `self.__class__.dlg = _HelpDialog(...)` is not too nice, so a reference might be better.

Comment (DietmarSchwertberger) [reply lib/matplotlib/backends/backend_wx.py]: Ah, `HelpWx.dlg = _HelpDialog(...)` would be an option.
This plus a comment would probably be the the best solution.

**Comment by anntzer** [comment]
Shouldn't the dialog just be made modal, and this way you don't have to handle lifetime problems?  (it is modal in gtk3 and qt5, not in tk right now)

**Comment by DietmarSchwertberger** [comment]
That would have been the easiest option, but that's not nice from a user perspective.
With this implementation the user can keep it open as long as he wants or just close it with Enter/ESC or the mouse.

**Comment by anntzer** [comment]
Does wx require you to keep an explicit reference to the dialog to avoid it getting garbage collected?
If that's the case, my solution (for qt, but the idea is the same) is to just have a global `_keep_alive` set and stash everything there (as there's typically no point of reimplementing that again and again).

**Comment by DietmarSchwertberger** [comment]
The two spaces are there, so the checker should not complain.

At first, I did not re-use, but then I sometimes ended with multiple dialogs open.
It`s almost over-engineering...

**Comment by timhoffm** [comment]
You could make the dialog sort of a singleton: Store the reference in the Dialog class and use a class method to instantiate the dialog.

**Comment by anntzer** [comment]
As @DietmarSchwertberger says we're very close to overengineering now :)
Would prefer a solution that works even if tools get added or removed (not necessarily common, but given that it shouldn't be hard to implement that...).

**Comment by anntzer** [comment]
Oh I didn't mean update on the fly, just update after close/open indeed.

**Comment by DietmarSchwertberger** [comment]
I like the singleton plus classmethod.
Updates on the fly would now be easy, as `_get_help_entries()` is called anyway, but without it's more readable.

File: lib/matplotlib/backends/backend_wx.py [hunk 5]
```diff
@@ -1826,9 +1826,61 @@ def remove_rubberband(self, dc=None):
...
+        OK = wx.Button(self, wx.ID_OK)
+        sizer.Add(OK, 0, wx.ALIGN_CENTER_HORIZONTAL | wx.ALL, 8)
+        self.SetSizer(sizer)
+        sizer.Fit(self)
+        self.Layout()
+        self.Bind(wx.EVT_CLOSE, self.OnClose)
+        OK.Bind(wx.EVT_BUTTON, self.OnClose)
+
+    def OnClose(self, evt):
+        _HelpDialog.instance = None  # remove global reference
```
[the remark this answers is not in the document]
Comment (DietmarSchwertberger) [reply lib/matplotlib/backends/backend_wx.py]: Oh, yes. You're right. Thanks.
(Worked anyway due to the bool value of a destroyed window, but kept an unneccesary Python reference.)

**Pull request merged**